### DISCUSSION TITLE
Interaction between "exclude" and LSP suggestions

### DISCUSSION THREAD

**Discussion by travisdowns** [microsoft/pyright#7984] [Q&A]
I'm creating a pyrightconfig.json for our moderately sized python project. This configuration will be used by CI to enforce that all included files type check cleanly. Since getting the entirely codebase clean is a herculean task I'm doing it incrementally: I `include` `**/*.py` but then `exclude` (in the config) a list of specific files that currently do not type check. I have tooling which detects if an excluded file actually became clean and so it can be removed.

What I noticed, however, is that putting a file in the "exclude" array in pyrightconfig.json prevents classes it contains from being auto-suggested as completions in other files (in vscode), prevents it from showing up in the "all symbols" listing, etc. I.e., it is mostly ignored by the LSP.

Is this by design? I just want to exclude these files form being part of the type set checked when I run pyright but I don't want to exclude them from the LSP entirely! It is seems clear that they are actually analyzed as if I use classes from the excluded I get analysis for them (e.g., if I use a non-existing attribute I am warned about it). If I open an excluded file in the editor it is type checked.

Is there any workaround here? I suppose I could maintain a different pyright config specifically for CLI use in CI, but nearly all of the other settings (e.g., which checks are enabled) I do want to share for consistency between IDE use and CLI use.

**Comment by erictraut** [comment]
Yes, this is by design. If you exclude a file, you're telling pyright (or pylance) that the file shouldn't be considered part of your project. It will no longer analyze that file (unless it happens to be imported by one of the files in your project), and it will no longer include that file in operations like auto-import or project-wide replace.

The "ignore" configuration setting may solve your problem. With this option, the files are still considered "part of the project" and are still analyzed, but all diagnostics for those files are suppressed. Does that meet your needs?

You may also be interested in [this feature request](https://github.com/microsoft/pyright/issues/4366) which I just opened for reconsideration. If this is something you're interested in, please upvote it.

**Comment by travisdowns** [comment]
Thanks Eric for your reply, it is much appreciated.

> The "ignore" configuration setting may solve your problem. With this option, the files are still considered "part of the project" and are still analyzed, but all diagnostics for those files are suppressed. Does that meet your needs?

Huh, the [documentation](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) makes it seem the other way around: exclude removes the file from analysis _except if it's included by an included flie_ while ignore removes the file from analysis _even if_ it's included?

I tested it and you are right, however. "Ignore" makes the symbols of the file show up in the LSP, whereas with "exclude" they are missing.

> It will no longer analyze that file (unless it happens to be imported by one of the files in your project), and it will no longer include that file in operations like auto-import or project-wide replace.

It does analyze it however. When I open any excluded file it is analyzed and errors show as usual, even if it is not referenced from any included file. However, if I add the file to `ignore` it is not analyzed. 

So I find the current behavior very confusing since neither option is really a superset of the other: `exclude` removes symbols from the LSP (regardless if referenced by other files or not) but still does type checking in the IDE when the file is opened. `ignore` is the reverse: symbols still show in the LSP but the file is unchcked when opened in the IDE.

**Reply by travisdowns** [reply]
I would add that in the vscode scenarios there are also many `python.` settings related to analysis and include/excludes that can be set and it is not clear how those interact with the pyrightconfig ones.

**Reply by erictraut** [reply]
I agree that the documentation for "include" and "exclude" isn't as clear as it could be. I've made some modifications that should help clarify.

The "include", "exclude" and "ignore" options have caused confusion with some users, but it's unlikely that we'd change them at this point. They were borrowed directly from the TypeScript compiler, which is similar to pyright in many ways.

> When I open any excluded file it is analyzed and errors show as usual

This is true of any python file that you open in your editor, even if it's completely outside of your project. This is useful, for example, when opening modules that are installed in the site-packages directory of your Python environment. These files are not considered part of your project and do not participate in project-wide operations like rename or call hierarchy. Files that you open in the editor that are not part of your project workspace are analyzed using a "default workspace" that has default type checker settings.

>  it is not clear how [language server settings] interact with the pyrightconfig ones

Most of the language server settings are specific to language server features and do not affect type checking. In cases where the language server settings do affect type checking behavior, the configuration takes precedence. For example, if your config file indicates that "typeCheckingMode" is "strict" for your project but your "python.analysis.typeCheckingMode" language server setting is set to "basic", the "strict" setting prevails. The assumption is that the config file is checked in to the repo and is common to all developers on the project, so these settings should override any per-developer preferences.

**Reply by travisdowns** [reply]
Thanks for your response Eric.

> This is true of any python file that you open in your editor

Yes, but then why are "ignore" files not analyzed in this way if opened in the IDE? Ignore is both stronger and weaker than exclude which seems to be the odd part.

So I if I understand it correctly "excluded" files are *not* part of the project which means they should be excluded from "top level" analysis though they maybe analyzed as part of analysis for files that are included (but even in this case, errors are not reported for this files: this is only so that errors can be reported in the included files possibly based on type information extracted from excluded files). When opened in the IDE they are analyzed using a default workspace. Since they are not part of the project they do not participate in the global workspace state like rename, call hierarchy find symbols, etc.

Then "ignored" files *are part of the project*, which means they _do_ participate in global operations like rename, call hierarchy, but errors are not flagged even when you open those files in the IDE.

> Files that you open in the editor that are not part of your project workspace are analyzed using a "default workspace" that has default type checker settings.

Thanks, this is useful info. So this will ignore settings in pyrightconfig.json but use settings set in vscode configuration like `python.foo` or will it also even ignore those latter settings?
